miniapps/dpg/convection-diffusion.cpp

```diff
@@ -0,0 +1,571 @@
+// |   |           |           |           |          |         |
+// | τ | (u ,∇⋅τ)  | 1/ε(σ , τ)|  <û,τ⋅n>  |          |    0    |
+
+// where (v,τ) ∈  H¹(Ωₕ) × H(div,Ωₕ)
+
+#include "mfem.hpp"
+#include "util/weakform.hpp"
+#include "../common/mfem-common.hpp"
+#include <fstream>
+#include <iostream>
+
+using namespace std;
+using namespace mfem;
+using namespace mfem::common;
+
+enum prob_type
+{
+   manufactured,
+   EJ
+};
+
+prob_type prob;
+Vector beta;
+double epsilon;
+// Function returns the solution u, and gradient du and the Laplacian d2u
+void solution(const Vector & x, double & u, Vector & du, double & d2u);
+double exact_u(const Vector & X);
+void exact_sigma(const Vector & X, Vector & sigma);
+double exact_hatu(const Vector & X);
+void exact_hatf(const Vector & X, Vector & hatf);
+double f_exact(const Vector & X);
+void setup_test_norm_coeffs(GridFunction & c1_gf, GridFunction & c2_gf);
+
+int main(int argc, char *argv[])
+{
+   // 1. Parse command-line options.
+   const char *mesh_file = "../../data/inline-quad.mesh";
+   int order = 1;
+   int delta_order = 1;
+   int ref = 1;
+   bool visualization = true;
+   int iprob = 0;
+   double theta = 0.0;
+   bool static_cond = false;
+   epsilon = 1e0;
+
+   OptionsParser args(argc, argv);
+   args.AddOption(&mesh_file, "-m", "--mesh",
+                  "Mesh file to use.");
+   args.AddOption(&order, "-o", "--order",
+                  "Finite element order (polynomial degree).");
+   args.AddOption(&delta_order, "-do", "--delta_order",
+                  "Order enrichment for DPG test space.");
+   args.AddOption(&epsilon, "-eps", "--epsilon",
+                  "Epsilon coefficient");
+   args.AddOption(&ref, "-ref", "--num_refinements",
+                  "Number of uniform refinements");
+   args.AddOption(&theta, "-theta", "--theta",
+                  "Theta parameter for AMR");
+   args.AddOption(&iprob, "-prob", "--problem", "Problem case"
+                  " 0: manufactured, 1: Erickson-Johnson ");
+   args.AddOption(&beta, "-beta", "--beta",
+                  "Vector Coefficient beta");
+   args.AddOption(&static_cond, "-sc", "--static-condensation", "-no-sc",
+                  "--no-static-condensation", "Enable static condensation.");
+   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
+                  "--no-visualization",
+                  "Enable or disable GLVis visualization.");
+   args.Parse();
+   if (!args.Good())
+   {
+      args.PrintUsage(cout);
+      return 1;
+   }
+   args.PrintOptions(cout);
+
+   if (iprob > 1) { iprob = 1; }
+   prob = (prob_type)iprob;
+
+   if (prob == prob_type::EJ)
+   {
+      mesh_file = "../../data/inline-quad.mesh";
+   }
+
+   Mesh mesh(mesh_file, 1, 1);
+   int dim = mesh.Dimension();
+
+   if (beta.Size() == 0)
+   {
+      beta.SetSize(dim);
+      beta = 0.0;
+      beta[0] = 1.;
+   }
+
+   // Define spaces
+   enum TrialSpace
+   {
+      u_space,
```

Comment:
Maybe I'm off base, but it might improve clarity to set these enum values = to their integers. E.G. e_space=0,sigma_space=1...  I say this because the methods you are feeding these to are expecting the integers

Reply:
done!